src/shm/buffer/zshm.rs

```diff
@@ -0,0 +1,127 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use std::{
+    borrow::{Borrow, BorrowMut},
+    mem::MaybeUninit,
+};
+
+use zenoh::shm::{zshm, zshmmut, ZShm};
+
+use crate::{
+    transmute::{
+        unwrap_ref_unchecked, unwrap_ref_unchecked_mut, Inplace, TransmuteFromHandle,
+        TransmuteIntoHandle, TransmuteRef, TransmuteUninitPtr,
+    },
+    z_loaned_shm_mut_t, z_loaned_shm_t, z_owned_shm_mut_t, z_owned_shm_t,
+};
+
+decl_transmute_owned!(Option<ZShm>, z_owned_shm_t);
+
+decl_transmute_handle!(zshm, z_loaned_shm_t);
+
+/// Constructs ZShm slice from ZShmMut slice
+#[no_mangle]
+pub extern "C" fn z_shm_from_mut(
+    this: *mut MaybeUninit<z_owned_shm_t>,
+    that: &mut z_owned_shm_mut_t,
+) {
+    let shm: Option<ZShm> = that.transmute_mut().extract().map(|val| val.into());
+    Inplace::init(this.transmute_uninit_ptr(), shm);
+}
+
+/// Constructs ZShm slice in its gravestone value.
+#[no_mangle]
+pub extern "C" fn z_shm_null(this: *mut MaybeUninit<z_owned_shm_t>) {
+    Inplace::empty(this.transmute_uninit_ptr());
+}
+
+/// Returns ``true`` if `this` is valid.
+#[no_mangle]
+pub extern "C" fn z_shm_check(this: &z_owned_shm_t) -> bool {
+    this.transmute_ref().is_some()
+}
+
+/// Converts borrowed ZShm slice as owned ZShm slice by performing shared memory handle copy
+#[no_mangle]
+pub extern "C" fn z_shm_copy(this: *mut MaybeUninit<z_owned_shm_t>, loaned: &z_loaned_shm_t) {
```

Comment:
Should we change the order of arguments here ? I.e. in terms of C++ is it considered as copy constructor (Shm(const Shm& other) ) or a class method like Shm clone() const ?

Reply:
I didn't find any copy methods like this, what is the convention for such functions?

Reply:
Also, is it like a deep copy of data, i.e. there will never be any observable effects on loaned if we modify this, or not (in this case it should probably be called clone) ?

Reply:
Neither of this. Under the hood this is a shallow copy, but shared memory buffers are very robust and provide inter-process write access safety. In other words, you will never run into seeing a side-effect in your z_*_shm_t as nobody elsewhere (intra- or inter process) will be able to construct z_*_shm_mut_t on this buffer. The only condition to get write access to SHM buffer is it's reference uniqueness!

Reply:
Treat this as a copy ctor. BUT! A copy ctor for SHM buffer - that will never do deep copy in terms of deep copy.